plugins/c-ares-dns-client.c

```diff
@@ -0,0 +1,757 @@
+		write_fd = watch->fd;
+
+	/* Give control to c-ares. */
+	ares_process_fd(watch->request->channel, read_fd, write_fd);
+
+	reset_ares_timeout(watch->request, destroy_old_source);
+
+	return TRUE;
+}
+
+/*
+ * Destroy an ares_watch structure. We register this as our value destroy
+ * function when creating the ares_watches table, and it is called by glib
+ * whenever we remove a value from the table or destroy the table.
+ */
+static void destroy_ares_watch(gpointer data)
+{
+	struct ares_watch *watch = data;
+
+	DBG("watch %p (fd %d)", watch, watch->fd);
+
+	if (watch->g_source_id != 0) {
+		if (!g_source_remove(watch->g_source_id)) {
+			DBG("g_source_remove failed for id %d",
+					watch->g_source_id);
+		}
+		watch->g_source_id = 0;
+	}
+
+	g_io_channel_unref(watch->gio_channel);
+	g_free(watch);
+}
+
+/*
+ * Create an ares_watch for |fd| and store it in the ares_watches table for
+ * |request|. Monitor for readability if |read| is TRUE. Monitor for writability
+ * if |write| is TRUE. If there is already an entry for |fd| in the table,
+ * update it according to the values of |read| and |write|.
+ */
+static gboolean init_ares_watch(struct ares_request *request, int fd,
+				gboolean read, gboolean write)
+{
+	struct ares_watch *watch;
+
+	DBG("fd = %d, read = %d, write = %d", fd, read, write);
+
+	/*
+	 * If there's an old watch in the table, destroy it. We'll replace it
+	 * with a new one below if c-ares is still interested in this fd.
+	 */
+	if (g_hash_table_lookup(request->ares_watches, &fd) != NULL) {
+		/* This removal calls destroy_ares_watch on the old watch. */
+		g_hash_table_remove(request->ares_watches, &fd);
+	}
+
+	if (!read && !write)
+		return TRUE;
+
+	watch = g_malloc0(sizeof(struct ares_watch));
+
+	watch->request = request;
+	watch->fd = fd;
+	watch->g_source_id = 0;
+
+	watch->gio_condition = G_IO_NVAL | G_IO_HUP | G_IO_ERR;
+	if (read)
+		watch->gio_condition |= G_IO_IN;
+	if (write)
+		watch->gio_condition |= G_IO_OUT;
+
+	watch->gio_channel = g_io_channel_unix_new(fd);
+	if (watch->gio_channel == NULL) {
+		ofono_error("%s: could not create g_io_channel for fd %d",
+			      __func__, fd);
+		g_free(watch);
+		return FALSE;
+	}
+	g_io_channel_set_close_on_unref(watch->gio_channel, FALSE);
+
+	g_hash_table_insert(request->ares_watches, &fd, watch);
+
+	watch->g_source_id = g_io_add_watch(watch->gio_channel,
+					    watch->gio_condition,
+					    ares_watch_io_cb,
+					    watch);
+
+	return TRUE;
+}
+
+/*
+ * Destroy an ares_request struct, freeing the resources allocated in
+ * init_ares_request. |request| must already have been removed from the
+ * |pending_requests| list and must have been marked not running.
+ */
+static void destroy_ares_request(struct ares_request *request)
+{
+	struct ares_addr_node *node, *next;
+
+	DBG("request %p", request);
+
+	ares_destroy(request->channel);
+	g_free(request->hostname);
+	if (request->interface) {
+		rp_filter_set(request->interface, TRUE);
```

Comment:
It's mentioned above that rp_filter_set doesn't really work...  As it effects the interface as a whole, should we keep this code in place if it's questionable whether or not it works?